### DISCUSSION TITLE
ANN: ChatGPT integration via any chatroom in cocalc

### DISCUSSION THREAD

**Discussion by williamstein** [sagemathinc/cocalc#6543] [Announcements]
See:

<img width="866" alt="image" src="https://user-images.githubusercontent.com/1276278/224513771-066bedba-daa1-4a29-a408-ff578881910e.png">

**Comment by williamstein** [comment]
<img width="1299" alt="image" src="https://user-images.githubusercontent.com/1276278/224834035-b0c0b68c-1302-4201-98c7-87f9ef793979.png">

**Comment by mforbes** [comment]
Can you integrate similarly with Bing AI?  I recall a discussion about ChatGPT 4 somewhere (can't find right now) but one mustt pay for this.  ChatGPT4 is is accessible through Bing AI for free, however, (at least for now), and is significantly better.  (Especially with providing attribution and sources.)

The following discussion about using these tools for teaching is quite good:

https://hbsp.harvard.edu/webinars/unlocking-the-power-of-ai/

**Reply by williamstein** [reply]
> Can you integrate similarly with Bing AI?

I just searched for a while, and despite a lot of random announcements and articles and bing chat itself saying "YES", as far as I can tell the answer is actually "NO right now, but someday it'll be an option.   There are some shady web-scraping approaches that will just result in my account getting banned, so obviously those aren't an option either.    For example, bing chat says "Microsoft has opened up Bing Chat with an API for developers ", but going to [the linked page ](https://www.microsoft.com/en-us/bing/apis) does not have a bing chat api. 

> I recall a discussion about ChatGPT 4 somewhere (can't find right now) but one must pay for this. 

Hopefully within a week, we will make ChatGPT 4 available on Cocalc.  (It's already available to us admins.)  It won't be free, since it costs 20x (!) as much as GPT-3.5; instead, it'll be on a "pay for what you use" model, i.e., each chat will cost some amount, you can see the total so far, and you'll get billed at the end of the month.  

The exact same answers apply to Google's Bard (and Anthropic), where the announcements about api access are vaporware. 

I do have access to https://cohere.com/'s LLM API, but it doesn't have a chat mode, and seems to me to be more expensive than GPT-3.5 and not quite as good, so not worth integrating for our use cases.

>  (Especially with providing attribution and sources.)

Just to clarify, GPT4 does not provide attribution or sources; it's bing that builds that on top. Also, GPT4 is about 4x slower than GPT-3.5, which can make a big difference.  For some things GPT-3.5 is a much better experience (due to speed or the task being easy, e.g., asking anything about git or bash, or fixing basic errors in python code), and for others GPT4 is much better (e.g., trickier reasoning problems, more complicated coding tasks).  

I did just add "neural search" (see https://github.com/sagemathinc/cocalc/pull/6692) to cocalc, which is the foundation for adding local knowledge to LLM's to make their output much more aware of information not in the original model, such as the latest python library docs, files in your own project, etc.   https://www.phind.com/ is another example of this built on gpt-4 or gpt-3.5, though only for public docs (so not your own files).

> https://hbsp.harvard.edu/webinars/unlocking-the-power-of-ai/

Thanks.
